### DISCUSSION TITLE
Add bluetooth battery module

### DISCUSSION THREAD

**Discussion by mahaveergurjar** [prasanthrangan/hyprdots#757] [Ideas]
I think Adding a feature to show Bluetooth headphone and etc devices' battery show on waybar makes this type of module
show battery at this time but in notification only when device is connecting
![240119_21h20m08s_screenshot](https://github.com/prasanthrangan/hyprdots/assets/57872034/2dfcfc2c-567a-4664-b701-f7d802444311)

but not show on way so add this feature 
![240119_21h22m40s_screenshot](https://github.com/prasanthrangan/hyprdots/assets/57872034/9844041e-2b2b-4944-8dee-0e0bc48893ec)

**Comment by kRHYME7** [comment]
Wow, a nice feature.is there any way to fetch the battery info in the terminal?

**Comment by mahaveergurjar** [comment]
yes 
first type bluetoothctl and then type info in and then you can see battery in last line

like this type
![240122_08h42m50s_screenshot](https://github.com/prasanthrangan/hyprdots/assets/57872034/0c028eb3-66bd-4a8d-98ac-f6a3de1b0d95)

**Reply by kRHYME7** [reply]
Nice can you provide the raw test output please I'll try to do some testing. So this is assuming only 1 device is connected right?

**Reply by mahaveergurjar** [reply]
Yes at this time only one device is connected 
you can make scripts so i can tested you script

**Reply by kRHYME7** [reply]
like this ? 

![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/f9d9afad-a1af-4d78-84cf-1d692d91cb2e)

**Reply by mahaveergurjar** [reply]
> like this ? 
> 
> ![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/f9d9afad-a1af-4d78-84cf-1d692d91cb2e)
> 
> 
> 

Yes this type 👍

**Comment by mahaveergurjar** [comment]
now connect two devices but show only the above device info by default 

![240122_09h40m28s_screenshot](https://github.com/prasanthrangan/hyprdots/assets/57872034/1082d0fe-38e3-43c5-ad3b-02ff17679c12)

using device address access any device
![240122_09h47m39s_screenshot](https://github.com/prasanthrangan/hyprdots/assets/57872034/09f9102d-0d26-4122-925d-a67936a66e8d)

**Reply by kRHYME7** [reply]
no worries this is already a waybar function but said to be experimental 

In case a single device without a battery will display only the number of connections ( this is the expected behaviour of the repo)
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/22c8ad49-5902-41a2-9448-dec791f07f59)


In case if  multiple device is connected ( Only tried on a single device with battery battery )
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/61aef868-7742-4890-b3e7-14b059b98643)

what else can we display if possible? volume is handled already by the other module. Is thi?s enough ?

**Comment by mahaveergurjar** [comment]
Yes this is enough

**Reply by kRHYME7** [reply]
do you have multiple Bluetooth devices with battery ? I do not know yet what will show if multiple devices have bluetooth

**Reply by mahaveergurjar** [reply]
Yes I have multiple devices with Bluetooth battery

**Reply by kRHYME7** [reply]
Here you go    
```
"bluetooth: {
        "format": "",
        "format-disabled": "",
        "format-connected": " {num_connections}",
        "format-connected-battery": " {device_alias} {device_battery_percentage}󰥉",
        "format-device-preference": [ "device2", "device1" ], // preference list deciding the displayed device
        "tooltip-format": "{controller_alias}\n{num_connections} connected",
        "tooltip-format-connected": "{controller_alias}\n{num_connections} connected\n\n{device_enumerate}",
        "tooltip-format-enumerate-connected": "{device_alias}",
        "tooltip-format-enumerate-connected-battery": "{device_alias}\t{device_battery_percentage}%"
        },

```
remember Some device won't declare there battery so sometimes it won't work and may look like this 
PinkBuds should have battery but dbus does not detect the battery percentage and same goes the bluetoothctl not fetching the percentage sometimes
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/8642d431-9c60-4653-9d8e-aa8f967f17a3)
bluetoothctl : Sometime not detecting the battery
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/bbdc72d3-7e84-4675-bbe1-ec6a3d927c02)

This  might be the reason why this is an experimental feature on waybar.

**Reply by mahaveergurjar** [reply]
Ok I will try

**Reply by kRHYME7** [reply]
Can you test connecting to those devices and try what device will be prioritized? 

BTW should I add a PR for this ?

**Reply by mahaveergurjar** [reply]
OK i will test

**Reply by kRHYME7** [reply]
Additional info: 

if you will see a notification plus battery % it is likely to have battery percentage in the waybar too. 
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/c92844da-284b-4633-97dc-7b1b16e1743a)
Waybar just reads the dbus or whatever that is called too, the same as how Dunst or a notification daemon knows the percentage. 
 So yes it will be hit and missed. but I also love this feature

**Comment by mahaveergurjar** [comment]
Yes this will work with both devices 
Yes please add PR for this waybar module 

![240122_10h49m24s_screenshot](https://github.com/prasanthrangan/hyprdots/assets/57872034/9b4a5760-69a5-4598-b186-60455a4ef3fd)

**Comment by mahaveergurjar** [comment]
I think you can add this line this type so see the batter in waybar
"format-connected-battery": " {device_alias} // 󰥉 {device_battery_percentage}%",
![240122_11h04m20s_screenshot](https://github.com/prasanthrangan/hyprdots/assets/57872034/345ace66-b546-469e-8260-1e1a52ff96be)

**Reply by kRHYME7** [reply]
I remove % as it looks too long but yes I might consider this as this is intuitive.

**Reply by mahaveergurjar** [reply]
ok

**Reply by kRHYME7** [reply]
How about this ? 

![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/13ea78bd-bf60-4a7f-a9b5-c98c45fa7aef)
        "format-connected-battery": "󰥉{device_alias} {device_battery_percentage}%",

**Reply by mahaveergurjar** [reply]
yeah its better

**Comment by mahaveergurjar** [comment]


**Reply by kRHYME7** [reply]
Make new

Relating to bluetooth batt ?

**Reply by kRHYME7** [reply]
Try to post this on a separate discussion so we can just throw this link to new comers and also easy tracking.
